pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,424 @@
+	logger := log.WithFields(log.Fields{"controller": controllerName})
+
+	if infraStatus.PlatformStatus == nil || infraStatus.PlatformStatus.AWS == nil {
+		return fmt.Errorf("Infrastructure platform status is not set")
+	}
+
+	r := &s3EndpointReconciler{
+		kubeclientset:      kubeclientset,
+		configclientset:    configclientset,
+		logger:             logger,
+		eventRecorder:      eventRecorder,
+		infrastructureName: infraStatus.InfrastructureName,
+		region:             infraStatus.PlatformStatus.AWS.Region,
+	}
+
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return isServiceAccountTokenSigner(e.MetaNew)
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return isServiceAccountTokenSigner(e.Meta)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Deleting the token signer configmap shouldn't be done.
+			// Be safe and don't react.  When the openshift-kube-apiserver-operator
+			// recreates it, we'll catch it.
+			return false
+		},
+	}
+
+	// Create a namespace local cache separate from the Manager cache
+	// A namespace scoped cache can still handle cluster scoped resources
+	cache, err := cache.New(config, cache.Options{Namespace: tokenSignerConfigMapNamespace})
+	if err != nil {
+		return err
+	}
+
+	informer, err := cache.GetInformer(context.TODO(), &corev1.ConfigMap{})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Informer{Informer: informer}, &handler.EnqueueRequestForObject{}, p)
+	if err != nil {
+		return err
+	}
+
+	mgr.Add(&oidcDiscoveryEndpointController{reconciler: r, cache: cache, logger: logger})
+
+	return nil
+}
+
+func isServiceAccountTokenSigner(meta metav1.Object) bool {
+	// all managed resources are named pod-identity-webhook
+	if meta.GetName() == tokenSignerConfigMapName {
+		return true
+	}
+	return false
+}
+
+type s3EndpointReconciler struct {
+	kubeclientset      *kubernetes.Clientset
+	configclientset    *configset.Clientset
+	logger             log.FieldLogger
+	eventRecorder      events.Recorder
+	infrastructureName string
+	region             string
+}
+
+var _ reconcile.Reconciler = &s3EndpointReconciler{}
+
+func (r *s3EndpointReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	r.logger.Info("reconciling AWS S3 OIDC discovery endpoint")
+
+	isUnmanaged, err := r.reconcileClusterAuthentication()
+	if err != nil {
+		r.logger.WithError(err).Error("failed reconciling cluster authentication")
+		return reconcile.Result{Requeue: true}, err
+	}
+	if isUnmanaged {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.reconcileS3Resources()
+	if err != nil {
+		r.logger.WithError(err).Error("failed reconciling S3 resources")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *s3EndpointReconciler) reconcileClusterAuthentication() (bool, error) {
```

Comment:
Also personally i would like if this function is named to hint what it actually does, like `reconcileServiceAccountIssuer`